### ISSUE TITLE
[BUG REPORT] Multiprocess pytorch dataloaders

### ISSUE THREAD

**Issue by rlizzo** [tensorwerk/hangar-py#135]
**Describe the bug**

The pytorch dataloader cannot currently be run with multiprocess workers:

```python
>>> torch_dset = make_torch_dataset(aset, index_range=slice(1, 100))
>>> loader = DataLoader(torch_dset, batch_size=16, num_workers=2)
>>> for batch in loader:
...     train_model(batch)
Exception: Cannot pickle `hangar.dataloaders.TorchDataset.BatchTuple`
```
This is because the the `BatchTuple` wrappers passed to `hangar.dataloaders.TorchDataset` are dynamically defined namedtuple classes, whose definition is not appropriately scoped for a forked subprocess to introspect it's name/contents upon pickling. 

https://github.com/tensorwerk/hangar-py/blob/e2c7a89ccb9ddb379e8a3fa8f20dae20fcfb6345/src/hangar/dataloaders/torchloader.py#L63-L74

As the structure needs to be dynamically defined based on user arguments, we cannot just place the `BatchTuple` definition in the main body of the module. 

### Two possible solutions:

@lantiga and @hhsecond, let me know what you prefer, or any other solutions you might have.

**1) keep the current return type exactally the same, but add the definition of `BatchTuple` to `globals()` before it is passed to `TorchLoader`**

```python
        wrapper = namedtuple('BatchTuple', field_names=field_names)
    else:
        wrapper = namedtuple('BatchTuple', field_names=gasets.arrayset_names, rename=True)

    globals()[`BatchTuple`] = wrapper
    return TorchDataset(gasets.arrayset_array, gasets.sample_names, wrapper)
```

   This works, but is generally bad practice to manually modify `global` scope. 

**2) return a `dict` of `field_names` and `tensors` instead of a `namedtuple`**

```python
        wrapper = tuple(field_names)
    else:
        wrapper = tuple(gasets.arrayset_names)

    return TorchDataset(gasets.arrayset_array, gasets.sample_names, wrapper)
```
And in `TorchDataset` replace:
https://github.com/tensorwerk/hangar-py/blob/e2c7a89ccb9ddb379e8a3fa8f20dae20fcfb6345/src/hangar/dataloaders/torchloader.py#L135
  with 
```python
    return dict(zip(self.wrapper, out))
```

  Which still works, and does not modify `globals()`, but changes the output o the function to something "not quite as nice" as a `namedtuple`.

**Severity**

Select an option:
- [ ] Data Corruption / Loss of Any Kind
- [ ] Unexpected Behavior, Exceptions or Error Thrown
- [x] Performance Bottleneck

**Comment by elistevens** [comment]
Make sure whatever option you pick works with having multiple `DataLoaders` in the main process, each loading something different.

**Comment by rlizzo** [comment]
Ahh excelent point. @lantiga, can you comment on this from a pytorch users perspective?

**Comment by elistevens** [comment]
Make the tuple class be named `'__'.join(['BatchTuple'] + field_names)`?

**Comment by lantiga** [comment]
Adding to `global` seems hacky to me. Adding field names as @elistevens would probably do it, but at that point the ability to unpickle depends on code execution.

Personally I'm ok with returning `dict` for now.

**Comment by elistevens** [comment]
Does returning a dict prevent the PyTorch special handling for named tuples
from working? Or does it handle dict collation too?

**Comment by lantiga** [comment]
Hey Eli. Can you make an example?

**Comment by elistevens** [comment]
```python
Python 3.6.7 (default, Oct 22 2018, 11:32:17)
[GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import collections
>>> NT = collections.namedtuple('NT', 'a,b,c')
>>> import torch
>>> torch.__version__
'1.2.0'
>>> import torch.utils.data
>>> class MyDataset(torch.utils.data.Dataset):
...     def __len__(self):
...         return 3
...     def __getitem__(self, ndx):
...         return NT(1, 2, 3)
...
>>> mds = MyDataset()
>>> mds[1]
NT(a=1, b=2, c=3)
>>> dl = torch.utils.data.DataLoader(mds, batch_size=2)
>>> l = list(dl)
>>> l[0]
NT(a=tensor([1, 1]), b=tensor([2, 2]), c=tensor([3, 3]))
```

**Comment by elistevens** [comment]
I SUPPOSE I could have checked that myself.  ;)

**Comment by rlizzo** [comment]
So consensus is `dict` then? I'll put a PR together in the next day or so.
